File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    def can_change(begin, target):
        diff_count = 0
        for i in range(len(begin)):
            if begin[i] != target[i]:
                diff_count += 1

        return diff_count == 1
    
    if target not in words:
        return 0
    
    queue = deque()
    queue.append((begin, 0))
    while queue:
        now_word, depth = queue.popleft()
        
        if now_word == target:
            return depth
        
        for w in words:
            if can_change(now_word, w):
                queue.append((w, depth + 1))

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int[] count = new int[words.length];
        boolean[] visited = new boolean[words.length];
        Queue<Object[]> queue = new LinkedList();
        queue.add(new Object[]{begin, 0});
        while (!queue.isEmpty()) {
            Object[] poll = queue.poll();
            String nowWord = (String) poll[0];
            int depth = (int) poll[1];
            
            if (nowWord.equals(target)) {
                return depth;
            }
            
            for (int i = 0; i < words.length; i++) {
                if (!visited[i] & canChange(nowWord, words[i])) {
                    visited[i] = true;
                    queue.add(new Object[]{words[i], depth + 1});
                }
            }
        }
        
        return 0;
    }
    
    public boolean canChange(String word, String target) {
        int count = 0;
        for (int i = 0; i < word.length(); i++) {
            if (word.charAt(i) != target.charAt(i)) {
                count++;
            }
        }
        
        return count == 1;
    }
}